### PR TITLE
New page on crossplane pods

### PR DESCRIPTION
**Pull request by plumbis**
Creates a new concept page covering the crossplane core pod and rbac manager pod.

### CODE REVIEW

File: content/master/concepts/pods.md
Comment (jeanduplessis) [inline content/master/concepts/pods.md:15]: Why is _init_ `P`od written with a capital here but in lowercase almost everywhere else?

Comment (jeanduplessis) [inline content/master/concepts/pods.md:15]: Why is _init_ `P`od written with a capital here but in lowercase almost everywhere else?

Comment (jeanduplessis) [inline content/master/concepts/pods.md:15]: Why is _init_ `P`od written with a capital here but in lowercase almost everywhere else?

File: content/master/concepts/pods.md
Comment (jeanduplessis) [inline content/master/concepts/pods.md:7]: Should two `P`ods be capitalized here?

File: content/master/concepts/pods.md
Comment (jeanduplessis) [inline content/master/concepts/pods.md:38]: What happens to the init pod after completion? Does the end user have to care about it at all after completion? Does it also run in the `crossplane-system` namespace?

File: content/master/concepts/pods.md
Comment (jeanduplessis) [inline content/master/concepts/pods.md:42]: It's getting confusing with the naming here. We refer to the main Crossplane pod as "core" pod, but its name is `crossplane` in the `crossplane-system` namespace, right? Not sure we can do any better than what's written but just calling it out.

File: content/master/concepts/pods.md
Comment (jeanduplessis) [inline content/master/concepts/pods.md:69]: Where should one change the polling rate? Can we be more specific than just stating to add it as a pod argument. Also is there any considerations that need to be taken into account when changing the poll-interval?

File: content/master/concepts/pods.md
Comment (jeanduplessis) [inline content/master/concepts/pods.md:126]: Why do we capitalize Composite Resources here, yet above we write managed resources in lowercase?

Comment (jeanduplessis) [inline content/master/concepts/pods.md:126]: Why do we capitalize Composite Resources here, yet above we write managed resources in lowercase?

File: content/master/concepts/pods.md
Comment (jeanduplessis) [inline content/master/concepts/pods.md:135]: Why/when would one want to switch from All to Basic policy?

File: content/master/concepts/pods.md
Comment (jeanduplessis) [inline content/master/concepts/pods.md:142]: Under what circumstances would you want to disable the RBAC manager?

File: content/master/concepts/pods.md
Comment (jeanduplessis) [inline content/master/concepts/pods.md:152]: Why/how would you run more than one Crossplane pod in a cluster?

File: content/master/concepts/pods.md
Comment (jeanduplessis) [inline content/master/concepts/pods.md:161]: Would this argument set which XP pod becomes the leader?

File: content/master/concepts/pods.md
Comment (jeanduplessis) [inline content/master/concepts/pods.md:176]: "Changing Crossplane pod..." should be "Change Crossplane pod..."

File: content/master/concepts/pods.md
Comment (jeanduplessis) [inline content/master/concepts/pods.md:204]: Line="7" doesn't seem like the right line here.

File: content/master/concepts/pods.md
Comment (jeanduplessis) [inline content/master/concepts/pods.md:208]: Line="10" doesn't seem like the right line here.

**Review by jeanduplessis — CHANGES_REQUESTED** [review]
@plumbis Thanks for starting this concept doc. I found it very valuable. Thanks for your patience with some of the noob questions on this content.

File: content/master/concepts/pods.md
Comment (plumbis) [reply content/master/concepts/pods.md:42]: More accurately it's the init _container_ and core _container_. Both of these containers are in the `crossplane` _pod_ inside the `crossplane-system` namespace. The name "core" is what Crossplane calls it in source. 

Would being more specific that they are two _containers_ inside a single _pod_ help? Otherwise we're stuck with

File: content/master/concepts/pods.md
Comment (bobh66) [reply content/master/concepts/pods.md:69]: The polling interval is an input to the helm chart for Crossplane deployment

File: content/master/concepts/pods.md
Comment (bobh66) [reply content/master/concepts/pods.md:15]: This should be "the init container" runs.  The Pod has "initContainers" that run at startup and "containers" that run after startup.

File: content/master/concepts/pods.md
Comment (bobh66) [reply content/master/concepts/pods.md:161]: No - it is random based on timing.  It is even possible that the "failed" pod could regain the leader role after it restarts if the timer for the "standby" pod(s) has not expired yet.

Comment (plumbis) [reply content/master/concepts/pods.md:161]: No, "who" is the leader doesn't matter since it's just another pod in the deployment. Just like the specific node a pod runs on doesn't matter. 

The election itself is driven by kubernetes. 

Should I add more context?

File: content/master/concepts/pods.md
Comment (plumbis) [reply content/master/concepts/pods.md]: fixed

File: content/master/concepts/pods.md
Comment (plumbis) [reply content/master/concepts/pods.md]: fixed.

File: content/master/concepts/pods.md
Comment (plumbis) [reply content/master/concepts/pods.md]: I added more information about viewing the init container in `k get pods` and that starts the core container automatically.

File: content/master/concepts/pods.md
Comment (plumbis) [reply content/master/concepts/pods.md]: fixed

File: content/master/concepts/pods.md
Comment (plumbis) [reply content/master/concepts/pods.md:170]: Added a note.

File: content/master/concepts/pods.md
Comment (plumbis) [reply content/master/concepts/pods.md:180]: Added a note that you can run more than one but most people don't need it. (Feedback from Nic)

File: content/master/concepts/pods.md
Comment (plumbis) [reply content/master/concepts/pods.md]: fixed

File: content/master/concepts/pods.md
Comment (plumbis) [reply content/master/concepts/pods.md]: fixed

File: content/master/concepts/pods.md
Comment (plumbis) [reply content/master/concepts/pods.md]: Fixed all pod/container references.

File: content/master/concepts/pods.md
Comment (plumbis) [reply content/master/concepts/pods.md:89]: I added some text on the impact of changing the poll-interval. 

@jeanduplessis are you okay with the note on "Read the Change Pod Settings section...." for information on applying the change or do you want more here?

File: content/master/concepts/pods.md
Comment (plumbis) [reply content/master/concepts/pods.md:91]: What would trigger an MR watch? I figured that MR changes would only happen based on external APIs, so it wouldn't be a notification. 

I wasn't able to figure out what is and isn't assigned a watch in source.

File: content/master/concepts/pods.md
Comment (plumbis) [reply content/master/concepts/pods.md:99]: Do you mean that it's 10 API calls per second AND 10 goroutines?  Meaning the value is two different settings?

File: content/master/concepts/pods.md
Comment (bobh66) [reply content/master/concepts/pods.md:102]: The MR controller is by definition "watching" all MRs of that type - it's the same for all Crossplane resource controllers, they all "watch" the specific composite/managed/etc resources that they are responsible for.  The synch interval is presumably intended to account for any "missed events" and verify that the resource state is what it should be regardless of whether the controller missed an event.  (This is my understanding - if Nic tells you something different, believe him.)

File: content/master/concepts/pods.md
Comment (bobh66) [reply content/master/concepts/pods.md:110]: Yes - as shown here:  https://github.com/crossplane/crossplane/blob/master/cmd/crossplane/core/core.go#L178

it is used for both MaxConcurrentReconciles (number of goroutines) and GlobalRateLimiter

File: content/master/concepts/pods.md
Comment (plumbis) [reply content/master/concepts/pods.md:102]: Insightful! Do you have an example of a MR event that would be caught by a watch?

Comment (bobh66) [reply content/master/concepts/pods.md:102]: Probably the simplest case is deletion - when someone or something deletes an MR kubernetes will add the `deletionTimestamp` and `deletionGracePeriodSeconds` attributes to the `metadata` section which triggers an event to the controller, the reconciler sees that the MR has been deleted and calls the external Delete() function to trigger resource deletion via the provider.  Another case would if the someone or something changed a value in the `spec` of the resource - that would also trigger an event to the controller which would detect the difference and trigger an update via the provider.

Note that polling/syncing and watched events are used together - the events are kubernetes-detected changes in (usually desired) state as above, where polling and syncing are looking for drift on the remote side (actual state).  Polling and syncing are time-based (check for differences every 5m/1h) where events are immediate (hey something changed you might care about - or not)

Comment (plumbis) [reply content/master/concepts/pods.md:103]: Thank you! I've added a MRs to the list of resources that uses watches as well as a notice under "MRs use polling" to explain when an MR would rely on the watch vs poll. 

I'm not trying to get too deep but just create awareness in the user.

File: content/master/concepts/pods.md
Comment (plumbis) [reply content/master/concepts/pods.md]: Thanks again. I added a bunch of stuff here to expand on it. I'm not 100% sure I got it right and accurate so I'd love a second review when you have a chance.

File: content/master/concepts/pods.md
Comment (negz) [inline content/master/concepts/pods.md:278]: Could we link to upstream Kubernetes documentation for this? This feels like we're teaching folks how to use Kubernetes.

**Review by negz — COMMENTED** [review]
I'm seeing a bunch of changes that don't seem to be related to the PR description - i.e. updates to content under the v1.12 and v1.9 directories. Is this intentional?

**Comment by plumbis** [comment]
> I'm seeing a bunch of changes that don't seem to be related to the PR description - i.e. updates to content under the v1.12 and v1.9 directories. Is this intentional?

I pulled in the v1.12 release either on accident or purpose. I've rebased and removed that commit. Only the pods.md file and vale updates are in the PR now.

File: content/master/concepts/pods.md
Comment (plumbis) [reply content/master/concepts/pods.md:34]: Do you mean why are some of them in code ticks?

File: content/master/concepts/pods.md
Comment (plumbis) [reply content/master/concepts/pods.md:70]: I don't think this should be moved, since the page flow is about the pods and their interactions/responsibilities but I do think the [root concepts page](https://docs.crossplane.io/v1.12/concepts/) cover this at or near the top instead of being a rehash of what's in the [Introduction](https://docs.crossplane.io/v1.12/getting-started/introduction/) page and what's spelled out in the individual concept sections.

File: content/master/concepts/pods.md
Comment (plumbis) [reply content/master/concepts/pods.md:111]: I added a line earlier "Some resources may be both watched and polled"

Comment (bobh66) [reply content/master/concepts/pods.md:111]: From the user's perspective it's essentially all resources that they care about/create are both watched and polled.

File: content/master/concepts/pods.md
Comment (plumbis) [reply content/master/concepts/pods.md:126]: Had those switched. Fixed.

File: content/master/concepts/pods.md
Comment (plumbis) [reply content/master/concepts/pods.md:142]: Gladly willing to change this. It was from trying to understand how the rate limiter works and ended up on Dan's [blog about client-go](https://danielmangum.com/posts/controller-runtime-client-go-rate-limiting/#the-default-controller-rate-limiter) describing the token bucket rate limiter it uses.

File: content/master/concepts/pods.md
Comment (plumbis) [reply content/master/concepts/pods.md:119]: I've attempted to simplify this section.

File: content/master/concepts/pods.md
Comment (plumbis) [reply content/master/concepts/pods.md:198]: >It's not clear to me what this means.

What isn't clear? I'm trying to indicate that the RBAC manager creates "some policies" that Crossplane needs to operate.

File: content/master/concepts/pods.md
Comment (plumbis) [reply content/master/concepts/pods.md:214]: I edited this section to hint at that. I'm trying to avoid fully describing every ClusterRole/Role that's created and the policies around it.

File: content/master/concepts/pods.md
Comment (plumbis) [reply content/master/concepts/pods.md:278]: The amount of "teaching kubernetes" is light and just provides practical context to applying the settings discussed earlier.

File: content/master/concepts/pods.md
Comment (negz) [reply content/master/concepts/pods.md:34]: In part, yes. Everything you've listed on the highlighted lines is both a Crossplane concept and a concrete API type. You could write any of them in three ways:

1. Composite Resource Definitions, Providers, Locks, Store Configs, etc.(Typical prose)
2. `CompositeResourceDefinition`, `Provider`, `Lock`, `StoreConfig`, etc. (`KindName` style)
3. `compositeresourcedefinitions`, `providers` `locks`, `storeconfigs`, etc. (`pluralname` style)

I would recommend using one of the first two styles consistently. The `pluralname` style where a type is written as lower case, plural, and in a code block is very uncommon relative to the `KindName` style.

File: content/master/concepts/pods.md
Comment (negz) [reply content/master/concepts/pods.md:70]: Yeah, I agree. It's not so much that I don't want it mentioned here insomuch as I think it should appear somewhere more prominent. If that were the case I could see this being a note on this page rather than an entire heading.

File: content/master/concepts/pods.md
Comment (negz) [reply content/master/concepts/pods.md:138]: Re-opening this thread as there's still some strange grammar in here, e.g.:

> When Crossplane checks a resource and it's not in the desired state __they__ 100 times per second.

Does my comment above help communicate what `--max-reconcile-rate` does? I'd like to make sure we remove any mention of anything happening 100 times per second - it's wrong and I think it would be worrying to readers. Let me know if I can help better explain how this works. Happy to take a stab at writing this section myself if necessary.

File: content/master/concepts/pods.md
Comment (negz) [inline content/master/concepts/pods.md:213]: I know you're attempting to be succinct here, but if you could briefly mention what the namespaced Crossplane RBAC roles are for I think it would help contextualize why folks would potentially use `Basic`.

Something like:

> When the RBAC manager uses the `All` policy it will create a series of opinionated RBAC roles in each namespace. You can bind these RBAC roles to users to grant them access to Crossplane types within a namespace.

File: content/master/concepts/pods.md
Comment (negz) [reply content/master/concepts/pods.md]: > What isn't clear? 

I think it's the missing "what for". The four broad things the RBAC manager does is:

1. Create and bind a role to Crossplane's service account, so it can control managed resources.
2. Create and bind roles to provider service accounts, so they can reconcile their managed resources.
3. Create cluster roles that you can bind to grant users access to Crossplane resources in all namespaces.
4. Create roles that you can bind to grant users access to Crossplane resources in a specific namespace.

Noting that 4 is disabled when you run the RBAC manager in `Basic` mode.

At the moment you capture the why of the first two well - these roles enable Crossplane and provider pods do what they need to do. What's throwing me off is that 3 and 4 are a bit over-simplified:

* I would briefly mention. what the roles are for.
* The roles grant access to claims, MRs, secrets, etc in addition to XRs. So I would suggest being non-specific and saying "Crossplane resources". Calling out XRs specifically implies to me they grant only that.

File: content/master/concepts/pods.md
Comment (negz) [reply content/master/concepts/pods.md]: Understood. I think it's worth documenting these policies in detail somewhere, but the "pods" page probably isn't it. Let's discuss in the thread at https://github.com/crossplane/docs/pull/412#discussion_r1183121100.

File: content/master/concepts/pods.md
Comment (plumbis) [reply content/master/concepts/pods.md:34]: After digging through the [k8s style guide](https://kubernetes.io/docs/contribute/style/style-guide/#use-upper-camel-case-for-api-objects), the [PR that pushed that guideline](https://github.com/kubernetes/website/pull/23497) and some [related discussion](https://github.com/kubernetes/website/issues/20862) the best answer seems to be the k8s style "[UpperCamelCase](https://en.wikipedia.org/wiki/Camel_case)"

The only problem I have with this is that it's confusing in how k8s describes when to use sentence-style capitalization and their docs are inconsistent in this styling so there aren't great examples to follow.

That being said I've updated this to use UpperCamelCase in the called out section and tried to update any other references in the rest of the doc.

File: content/master/concepts/pods.md
Comment (plumbis) [reply content/master/concepts/pods.md:76]: moved.

File: content/master/concepts/pods.md
Comment (plumbis) [reply content/master/concepts/pods.md:70]: I understand. Making it just a note doesn't feel like it's enough. I think the solution is we need to give reconciliation more attention in the parent /concepts page.

File: content/master/concepts/pods.md
Comment (plumbis) [reply content/master/concepts/pods.md:138]: I've simplified the wording here and removed the reference to 100x

File: content/master/concepts/pods.md
Comment (plumbis) [reply content/master/concepts/pods.md:143]: Removed

File: content/master/concepts/pods.md
Comment (plumbis) [reply content/master/concepts/pods.md]: I was confused in your earlier comments on the `Roles` vs `ClusterRoles`. I've rewritten the section and I chose not to detail the `Roles`. Maybe in the future we can create an RBAC section that would go into details.

File: content/master/concepts/pods.md
Comment (plumbis) [reply content/master/concepts/pods.md]: I was confused in your earlier comments on the `Roles` vs `ClusterRoles`. I've rewritten the section and I chose not to detail the Roles. Maybe in the future we can create an RBAC section that would go into details.

File: content/master/concepts/pods.md
Comment (negz) [inline content/master/concepts/pods.md:144]: I'd still prefer not to talk about threads. It's not accurate. Could we find some other approachable way to phrase this? i.e. How many resources can be reconciled concurrently?

File: content/master/concepts/pods.md
Comment (negz) [inline content/master/concepts/pods.md:232]:
```suggestion
kubectl describe clusterrole crossplane-admin
```

`crossplane:aggregate-to-admin` isn't the full role, it's just part of it. `crossplane-admin` is the full role.

(Ditto for the edit, view, and browse roles below.)

**Review by negz — COMMENTED** [review]
I've only reviewed the content in the `master` directory, not `v1.11` or `v1.12` which I presume are identical.

File: content/master/concepts/pods.md
Comment (plumbis) [reply content/master/concepts/pods.md:144]: I'm struggling with this. I understand why you want to avoid `thread` but it seems like we're hinting at something without actually naming it. How is this edit? 
---
The second value `--max-reconcile-rate` defines is the number of
resources that can reconcile at once. If there are more resources than
configured `--max-reconcile-rate` the remaining resources must wait until
Crossplane reconciles a an existing resource.

Read the [Change Pod Settings]({{<ref "#change-pod-settings">}}) section for
instructions on applying these settings.

Comment (negz) [reply content/master/concepts/pods.md]: That sounds great to me.

**Review by negz — APPROVED** [review]
This looks great, thanks for all the effort!

**Pull request merged**
